Framework/Algorithms/src/SolidAngle.cpp

```diff
@@ -26,6 +28,73 @@ DECLARE_ALGORITHM(SolidAngle)
 
 using namespace Kernel;
 using namespace API;
+using namespace Geometry;
+
+namespace {
+
+/**
+ * Returns the angle between the sample-to-pixel vector and its
+ * projection on the X-Z (vertical tube) or Y-Z (horizontal tube) plane.
+ * Note, in all cases Y is assumed to be the pointing-up direction, Z is the
+ * beam direction.
+ */
+double getTubeAngle(const DetectorInfo &detectorInfo, size_t index,
+                    const bool vertical) {
+  const auto sampleDetVec =
+      detectorInfo.position(index) - detectorInfo.samplePosition();
+  auto inPlane = sampleDetVec;
+  if (vertical)
+    inPlane.setY(0.0);
+  else
+    inPlane.setX(0.0);
+  return sampleDetVec.angle(inPlane);
+}
+
+/**
+ *Returns correct angular function for the specified method
+ */
+std::function<double(size_t)>
```

Comment:
You may want to further investigate the overhead from `std::function`. I think using templates would be faster, but also increase the amount of duplicate code.

Reply:
Current version runs in <0.1s. I'd have to see a case that warrants the investigation.

Reply:
@peterfpeterson In general I agree with you which I why I used it in our branch. @gvardany mentioned earlier that 0.1 sec vs 0.03 sec was a "non-negligible speed-up for SANS treatment," which makes me suspect this is being run a zillion times.

Reply:
I think the current performance is reasonable. Will try again with the current version where l2 is taken from the pixels, this might be slightly slower than caching the `dOmega(0)`. Will give it a go with template method pattern, that seems the best fit to avoid the overhead and the duplicated code simultaneously.